DTS-STN/secure-client-hub
# Update pages to handle session validation better

## [ADO-179388](https://dev.azure.com/VP-BD/DECD/_workitems/edit/179388)

Fixed [AB#179388](https://dev.azure.com/VP-BD/d5ca3cd4-19cb-4c8c-b3e6-a7068f4677e4/_workitems/edit/179388)

### Changelog
feat: add visibilitychange session validation
feat: add session validation to getServerSideProps to check validity on load 

### Description of proposed changes:
This PR updates all of our pages to include session validation on load and on visibility change (when changing tabs or opening a minimized window for example).

### What to test for/How to test
1. Pull in branch
2. Ensure auth is enabled
3. Type `npm run build` and `npm run start`
4. Login to SYS2/Stream1. Once logged in, open a new tab and navigate to http://localhost:3000 and select a language
5. Once redirected to dashboard, inspect the page and open the network tab
6. Try switching in between tabs. When swapping back to the dashboard, you should see a request to `/api/refresh-msca` with a 200 response. Try minimizing and maximizing the window and confirm the same request appears (after 15 seconds as the event is throttled)
7. Try refreshing the page and confirm the same request appears
8. With the dashboard page still open, go back to the MSCA home page tab and log out. When you go back to the dashboard tab you should be redirected to the logout page
9. Repeat step 8 but open a new tab and navigate to the dashboard instead

### Additional Notes
As of this morning, there have been some `503 Service Unavailable` errors when navigating to many of the MSCA pages on multiple streams (Stream 1 and Stream 4 were the only ones I had tried given those are the only ones we have accounts for). This will make testing the logout functionality a bit difficult until that is back up and running, but I was able to test it last night with relative success so hopefully it works later today.
